src/Roave/SecurityAdvisories/Advisory.php

```diff
@@ -56,11 +58,14 @@ private function __construct(string $componentName, array $branchConstraints)
     /**
      * @param string[]|string[][][]|string[][][][] $config
      *
+     * @return Advisory
+     *
      * @throws InvalidArgumentException
+     *
+     * @psalm-suppress RedundantCondition
```

Comment:
`assert(is_string($componentName))` can be dropped

Reply:
Oh, I see... is it just phpstan complaining? I'm thinking of dropping that completely, if that's an issue.

Psalm correctly infers that `str_replace :: string -> string -> string -> string`, while phpstan may be inferring `str_replace :: string -> string -> string -> string|[string]` here, if I understand it correctly

Reply:
yup, when I delete `assert(is_string($componentName));` phpstan outputs error
<details>
<summary>output</summary>


```sh
Note: Using configuration file /home/itribusean/personal/Projects/SecurityAdvisoriesBuilder/phpstan.neon.dist.
 24/24 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%                                      
                                                                                
 ------ ------------------------------------------------------------------------------------------------------------------------
  Line   src/Roave/SecurityAdvisories/Advisory.php                                                                              
 ------ ------------------------------------------------------------------------------------------------------------------------
  73     Parameter #1 $componentName of class Roave\SecurityAdvisories\Advisory constructor expects string, array|string given.
 ------ ------------------------------------------------------------------------------------------------------------------------
                                                                                
                                                                                                                        
 [ERROR] Found 1 error                                                                                                  
                                                                                                                        

```

</details>

All other tooling behaves - pipeline is green.

Reply:
I think this may be because of the older PHPStan version we're running: can you try upgrading all `phpstan/*` dependencies when you have some time/will?

Reply:
sorry I did not notice your last comment :cold_sweat:   
so I did an update for `phpstan/*`:
<details>
<summary>composer info before update</summary>

```sh
itribusean@Nebuchadnezzar ~/p/P/SecurityAdvisoriesBuilder (php7.4)> composer info|grep phpstan
phpstan/phpdoc-parser                          0.3.5   PHPDoc parser with support for nullable, intersection and generic types
phpstan/phpstan                                0.11.19 PHPStan - PHP Static Analysis Tool
phpstan/phpstan-phpunit                        0.11.2  PHPUnit extensions and rules for PHPStan
phpstan/phpstan-strict-rules                   0.11.1  Extra strict and opinionated rules for PHPStan
thecodingmachine/phpstan-safe-rule             v0.1.4  A PHPStan rule to detect safety issues. Must be used in conjunction with thecodingmachine/safe
```

</details>

<details>
<summary>composer info after update</summary>

```sh
itribusean@Nebuchadnezzar ~/p/P/SecurityAdvisoriesBuilder (php7.4)> composer info|grep phpstan
phpstan/phpdoc-parser                          0.4.3   PHPDoc parser with support for nullable, intersection and generic types
phpstan/phpstan                                0.12.11 PHPStan - PHP Static Analysis Tool
phpstan/phpstan-phpunit                        0.12.6  PHPUnit extensions and rules for PHPStan
phpstan/phpstan-strict-rules                   0.12.2  Extra strict and opinionated rules for PHPStan
thecodingmachine/phpstan-safe-rule             v1.0.0  A PHPStan rule to detect safety issues. Must be used in conjunction with thecodingmachine/safe
```

</details>

and here is what `phpstan` says now:
<details>
<summary> output: </summary>

```sh
itribusean@Nebuchadnezzar ~/p/P/SecurityAdvisoriesBuilder (php7.4)> vendor/bin/phpstan analyse

Note: Using configuration file /home/itribusean/personal/Projects/SecurityAdvisoriesBuilder/phpstan.neon.dist.
 24/24 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   build-conflicts.php                                                                                                                                                                                                              
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  67     Short ternary operator is not allowed. Use null coalesce operator if applicable or consider using long ternary.
  244    Anonymous function should have native return typehint "array".
  256    Class DateTime is unsafe to use. Its methods can return FALSE instead of throwing an exception. Please add 'use Safe\DateTime;' at the beginning of the file to use the variant provided by the 'thecodingmachine/safe' library.
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ ------------------------------------------------------------------------------------------------------------------------
  Line   src/Roave/SecurityAdvisories/Advisory.php                                                                              
 ------ ------------------------------------------------------------------------------------------------------------------------
  49     Short ternary operator is not allowed. Use null coalesce operator if applicable or consider using long ternary.
  71     Parameter #1 $componentName of class Roave\SecurityAdvisories\Advisory constructor expects string, array|string given.
  74     Anonymous function should have native return typehint "Roave\SecurityAdvisories\VersionConstraint".
  98     Short ternary operator is not allowed. Use null coalesce operator if applicable or consider using long ternary.
  101    Anonymous function should have native return typehint "string".
 ------ ------------------------------------------------------------------------------------------------------------------------

 ------ --------------------------------------------------------------------------------------------
  Line   src/Roave/SecurityAdvisories/AdvisorySources/GetAdvisoriesFromFriendsOfPhp.php             
 ------ --------------------------------------------------------------------------------------------
  54     Anonymous function should have native return typehint "Roave\SecurityAdvisories\Advisory".
  76     Anonymous function should have native return typehint "bool".
 ------ --------------------------------------------------------------------------------------------

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   src/Roave/SecurityAdvisories/AdvisorySources/GetAdvisoriesFromGithubApi.php                                                                                                                                
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  133    Method Roave\SecurityAdvisories\AdvisorySources\GetAdvisoriesFromGithubApi::getAdvisories() should return array<int, array<string, array<string, array<string, string>|string>|string>> but returns array.
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ -----------------------------------------------------------------
  Line   src/Roave/SecurityAdvisories/Component.php                      
 ------ -----------------------------------------------------------------
  60     Anonymous function should have native return typehint "string".
  66     Anonymous function should have native return typehint "array".
 ------ -----------------------------------------------------------------

 ------ -----------------------------------------------------------------
  Line   src/Roave/SecurityAdvisories/VersionConstraint.php              
 ------ -----------------------------------------------------------------
  80     Anonymous function should have native return typehint "string".
 ------ -----------------------------------------------------------------

 ------ --------------------------------------------------------------------------------------
  Line   test/RoaveTest/SecurityAdvisories/AdvisorySources/GetAdvisoriesFromGithubApiTest.php 
 ------ --------------------------------------------------------------------------------------
  126    Anonymous function should have native return typehint "bool".
 ------ --------------------------------------------------------------------------------------

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   test/RoaveTest/SecurityAdvisories/AdvisorySources/GetAdvisoriesFromMultipleSourcesTest.php                                                                                   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  52     Method RoaveTest\SecurityAdvisories\AdvisorySources\GetAdvisoriesFromMultipleSourcesTest::getGenerator() return type has no value type specified in iterable type Generator.
         💡 Consider adding something like Generator<Foo> to the PHPDoc.
         You can turn off this check by setting checkMissingIterableValueType: false in your phpstan.neon.dist.
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ -----------------------------------------------------------------
  Line   test/RoaveTest/SecurityAdvisories/VersionConstraintTest.php     
 ------ -----------------------------------------------------------------
  379    Anonymous function should have native return typehint "string".
  481    Anonymous function should have native return typehint "string".
  559    Anonymous function should have native return typehint "string".
  603    Anonymous function should have native return typehint "string".
 ------ -----------------------------------------------------------------

 ------ -----------------------------------------------------------------
  Line   test/RoaveTest/SecurityAdvisories/VersionTest.php               
 ------ -----------------------------------------------------------------
  243    Anonymous function should have native return typehint "string".
  316    Anonymous function should have native return typehint "string".
 ------ -----------------------------------------------------------------

                                                                                                                        
 [ERROR] Found 22 errors                                                                                                
                                                                                                                        

```
</details>

So seems like we need phpstan. If you will I'm going to leave it like that and probably update other dependencies and fix what is up. What do you think ?

Reply:
I'd say that we should remove all `phpstan/*` dependencies if psalm already works

Reply:
that is a bit scary to be honest. I really like `phpstan` and `codingmachine/phpstan-safe-rule`
I figured If level of strictness is decreased then we have green pipeline, like so:
<details>
<summary>diff</summary>

```diff
diff --git a/phpstan.neon.dist b/phpstan.neon.dist
index b040bc0..5b61f5b 100644
--- a/phpstan.neon.dist
+++ b/phpstan.neon.dist
@@ -1,5 +1,5 @@
 parameters:
-       level: 7
+       level: 5
        paths:
                - build-conflicts.php
                - public
diff --git a/src/Roave/SecurityAdvisories/Advisory.php b/src/Roave/SecurityAdvisories/Advisory.php
index a9a312c..057506c 100644
--- a/src/Roave/SecurityAdvisories/Advisory.php
+++ b/src/Roave/SecurityAdvisories/Advisory.php
@@ -26,7 +26,6 @@ use function array_values;
 use function assert;
 use function implode;
 use function is_array;
-use function is_string;
 use function Safe\usort;
 use function str_replace;
 
@@ -61,15 +60,12 @@ final class Advisory
      * @return Advisory
      *
      * @throws InvalidArgumentException
-     *
-     * @psalm-suppress RedundantCondition
      */
     public static function fromArrayData(array $config) : self
     {
         $componentName = str_replace('composer://', '', $config['reference']);
         $branches      = $config['branches'];
 
-        assert(is_string($componentName));
         assert(is_array($branches));
 
         return new self(

```
</details>
With level 5 unsafe functions will still be detected.
Maybe we can leave phpstan but make it less strict?

Reply:
Psalm is way more advanced than phpstan as it currently stands: don't be too scared to drop it, especially now that we start going into better typed territory